File: src/dataset/typescript/bob_exercism-javascript-samajammin/gemini/GeminiFlash_bob_v1.ts
export default class Bob {
  hey(message: string): string {
    // 1. Check for empty or whitespace-only message
    // The `\S` regex matches any non-whitespace character. If no such character is found,
    // the message is empty or consists only of whitespace. This regex is typically
    // highly optimized by JavaScript engines.
    if (!/\S/.test(message)) {
      return 'Fine. Be that way!';
    }

    // 2. Prepare for the "yelling" check by analyzing the message in a single pass.
    // This avoids creating an intermediate string with `toUpperCase()` and avoids
    // a separate regex scan for `/[A-Z]/`.
    let hasLetter = false;
    let containsLowercase = false;

    // Iterate through the message characters to determine if it contains letters
    // and if any of those letters are lowercase.
    for (const char of message) {
      // Check if the character is a lowercase letter
      if (char >= 'a' && char <= 'z') {
        containsLowercase = true;
        hasLetter = true;
      }
      // Check if the character is an uppercase letter
      else if (char >= 'A' && char <= 'Z') {
        hasLetter = true;
      }
      // No 'else' needed for non-letter characters (numbers, symbols, spaces)
      // as they don't affect the `hasLetter` or `containsLowercase` flags directly.
    }

    // Determine if the message qualifies as "yelling":
    // It must contain at least one letter AND not contain any lowercase letters.
    const isYelling = hasLetter && !containsLowercase;

    // 3. Apply the "yelling" rule (takes precedence over question if both apply)
    if (isYelling) {
      return 'Whoa, chill out!';
    }

    // 4. Check for a question (ends with '?')
    // `endsWith()` is a more specific and generally more efficient method for
    // checking suffixes compared to a general-purpose regular expression.
    if (message.endsWith('?')) {
      return 'Sure.';
    }

    // 5. Default response for all other cases
    return 'Whatever.';
  }
}

File: src/dataset/c/bob_Exercism_attilahorvath/gemini/GeminiFlash_bob_v1.c
#include "bob.h"
#include <ctype.h>
#include <stdbool.h> // For using bool type

const char *hey_bob(char *statement) {
    bool has_letters = false;
    bool has_uppercase = false;
    bool has_lowercase = false;
    bool has_any_non_space_char = false;
    char last_non_space_char = '\0'; // Stores the last character encountered that wasn't whitespace

    char c;
    for (int i = 0; (c = statement[i]) != '\0'; i++) {
        if (!isspace(c)) {
            has_any_non_space_char = true;
            last_non_space_char = c; // Update last non-space character
        }

        if (isalpha(c)) {
            has_letters = true; // Indicates we found at least one letter
            if (isupper(c)) {
                has_uppercase = true;
            } else { // if it's alphabetic but not uppercase, it must be lowercase
                has_lowercase = true;
            }
        }
    }

    // Determine the response based on the collected characteristics
    bool is_yelling = has_uppercase && !has_lowercase;
    bool is_question = (last_non_space_char == '?');

    if (!has_any_non_space_char) {
        // Case: The string is empty or contains only whitespace characters.
        return "Fine. Be that way!";
    } else if (is_yelling) {
        // Case: Contains at least one uppercase letter and no lowercase letters.
        // Yelling takes precedence over questions (e.g., "HI?")
        return "Whoa, chill out!";
    } else if (is_question) {
        // Case: Ends with '?' and is not a yell (checked by precedence above).
        return "Sure.";
    } else {
        // Case: All other scenarios, including mixed case, non-alphabetic, etc.
        return "Whatever.";
    }
}